tuxedo-rs: init at 0.2.2

mrcjkb opened the pull request:
###### Description of changes

This adds [tuxedo-rs](https://github.com/AaronErhardt/tuxedo-rs), which consists of:

* A dbus daemon as well as a simple CLI for interacting with the hardware of TUXEDO computers
* A GUI (tailor_gui), which can be used to manage power/keyboard-led/fan profiles.

tuxedo-rs is an alternative to the [TUXEDO Control Center (TCC)](https://www.tuxedocomputers.com/en/TUXEDO-Control-Center.tuxedo), for which there exists a packaging request: #132206.

Note that if the packaging request for TCC is fulfilled, the daemons will conflict with each other, so the respective modules should ensure that the other one is disabled.

I had previously contributed [a flake](https://github.com/AaronErhardt/tuxedo-rs/blob/main/flake.nix) to tuxedo-rs, on which I have based this PR - with a small change to the module (`options.services` -> `options.hardware`).
If this is merged, I plan to rewrite the module in the flake to use this one and override the package `src`.
I have been using the tuxedo-rs module on my device for a few weeks now.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
maintainers: add mrcjkb

geri1701 reviewed (changes requested):
Please change the PR title from `tuxedo-rs: intit at 0.2.1` to `tuxedo-rs: init at 0.2.1`.

NobbZ commented:
Please rebase and fix merge conflicts.

Also personally I'd prefer to split "init" of the package and module into 2 commits. I am not sure about waht the consensus is on that. Therefore not "demanding" or "requiring" any change on that. Just putting it to the discussion here.

mrcjkb commented:
> Please rebase and fix merge conflicts.
> 
> Also personally I'd prefer to split "init" of the package and module into 2 commits. I am not sure about waht the consensus is on that. Therefore not "demanding" or "requiring" any change on that. Just putting it to the discussion here.

I have rebased and split the tuxedo-rs commit into separate package init and a module init commits.

geri1701 commented on pkgs/os-specific/linux/tailor_gui/default.nix:
Please remove any punctuation from `meta.description` and please note that the value should only be one sentence long.

geri1701 reviewed (commented):
If I clone your branch, I can build both `tailor_gui` and `tuxedo-rs` without problems. But something is obviously wrong with the output paths here, I don't know what to do about it at the moment, someone with more experience should take a look at that.

mrcjkb commented:
> If I clone your branch, I can build both `tailor_gui` and `tuxedo-rs` without problems. But something is obviously wrong with the output paths here, I don't know what to do about it at the moment, someone with more experience should take a look at that.

Same here.
Thanks for the review 😄

mrcjkb commented:
It doesn't seem to like

```nix
inherit ((lib.importTOML "${src}/tailord/Cargo.toml").package) version;
```

and the use of `builtins.path`.

geri1701 commented:
> It doesn't seem to like
> 
> ```nix
> inherit ((lib.importTOML "${src}/tailord/Cargo.toml").package) version;
> ```

Yes, in the output of nixpkgs-review it looks like it doesn't like that part.

mrcjkb commented:
It seems like it was actually failing because of 
```nix
cargoDeps = rustPlatform.importCargoLock {
    lockFile = src + "/Cargo.lock";
};
```

`lockFile` has to be a path.
The latest changes are working on my local machine (tested with `nix-env -qaP --out-path -f default.nix --show-trace`).

geri1701 commented:
Result of `nixpkgs-review pr 231384` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

geri1701 commented on pkgs/os-specific/linux/tailor_gui/default.nix:
Please change `description = lib.literalMD "Rust GUI for interacting with hardware from TUXEDO Computers";` to `description = "Rust GUI for interacting with hardware from TUXEDO Computers";` and `longDescription = lib.mdDoc ''
      An alternative to the [TUXEDO Control Center](https://www.tuxedocomputers.com/en/TUXEDO-Control-Center.tuxedo),
      written in Rust.
    '';` to `longDescription =  ''An alternative to the TUXEDO Control Center  (https://www.tuxedocomputers.com/en/TUXEDO-Control-Center.tuxedo), written in Rust. '';`. In both cases, a string is expected here.

geri1701 commented on nixos/modules/services/hardware/tuxedo-rs.nix:
`systemd.services.<name>.description`  should be of type string.
Please change it to `description = "Tuxedo Tailor hardware control service";`

geri1701 commented:
It looks good now.

geri1701 commented:
Result of `nixpkgs-review pr 231384` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>tailor_gui</li>
    <li>tuxedo-rs</li>
  </ul>
</details>

blitz commented:
I've previously used the flake on my Tuxedo laptop with no issues. As the person who failed to properly package the Tuxedo Control Center Electron monster, I'm more than happy that this implementation lands in nixpkgs! :tada:

xaverdh reviewed (approved):
I can confirm that this works on my Tuxedo Aura Gen1, thanks for the effort!

gonzaloetjo commented:
Worked for me, thanks!

ners commented:
As you have added a new module, please mention it in the release notes.

mrcjkb commented:
> As you have added a new module, please mention it in the release notes.

Done.

mrcjkb commented:
@figsoda (sorry for the ping)
Is there anything in the way of getting this merged?

figsoda commented on pkgs/os-specific/linux/tailor_gui/default.nix:
```diff
@@ -0,0 +1,73 @@
+  ];
+
+  postFixup = ''
+    wrapProgram $out/bin/tailor_gui --set XDG_DATA_DIRS "$out/share/gsettings-schemas/tailor_gui"
```
maybe we should let `wrapGAppsHook4` handle this instead

figsoda commented on pkgs/os-specific/linux/tailor_gui/default.nix:
```diff
@@ -0,0 +1,73 @@
+      written in Rust.
+    '';
+    homepage = "https://github.com/AaronErhardt/tuxedo-rs";
+    license = licenses.gpl2Only;
```
[Cargo.toml](https://github.com/AaronErhardt/tuxedo-rs/blob/a1338dba111b422b5fc6dcb9c710b61a9594a191/tailord/Cargo.toml#L6) seems to suggest this

```suggestion
    license = licenses.gpl2Plus;
```

figsoda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -35898,6 +35898,8 @@ with pkgs;
 
   tailor = callPackage ../applications/version-management/tailor { };
 
+  tailor_gui = callPackage ../os-specific/linux/tailor_gui { };
```
```suggestion
  tailor-gui = callPackage ../os-specific/linux/tailor-gui { };
```

figsoda commented on pkgs/os-specific/linux/tuxedo-rs/default.nix:
```diff
@@ -0,0 +1,48 @@
+
+in
+rustPlatform.buildRustPackage {
+
+  pname = "tuxedo-rs";
+
+  inherit src;
+
+  # NOTE: OfBorg doesn't like importTOML from a subdirectory
+  version = "0.2.1";
```
`src` can probably be inlined here

```suggestion
  pname = "tuxedo-rs";
  version = "0.2.1";

  inherit src;
```

figsoda commented on pkgs/os-specific/linux/tuxedo-rs/default.nix:
```diff
@@ -0,0 +1,48 @@
+  cargoHash = "sha256-cmsHigW19LKT9Z0AVhmXKyB+Rzr++rG37UhkYxGaKTk=";
+
+  postInstall = ''
+    mkdir -p $out/share/dbus-1/system.d
+    cp ${src}/tailord/com.tux.Tailor.conf $out/share/dbus-1/system.d
```
```suggestion
    install -Dm444 tailord/com.tux.Tailor.conf -t $out/share/dbus-1/system.d
```

figsoda commented on pkgs/os-specific/linux/tuxedo-rs/default.nix:
```diff
@@ -0,0 +1,48 @@
+      - tailor: CLI
+    '';
+    homepage = "https://github.com/AaronErhardt/tuxedo-rs";
+    license = licenses.gpl2Only;
```
```suggestion
    license = licenses.gpl2Plus;
```

mrcjkb commented:
:pray:

figsoda commented on pkgs/os-specific/linux/tailor-gui/default.nix:
```diff
@@ -0,0 +1,69 @@
+    pkg-config
+    desktop-file-utils
+    appstream-glib
+    wrapGAppsHook4
```
this needs to be added to the lambda parameters

mrcjkb replied:
Oops...

mrcjkb commented:
```console
# x86_64-linux
> nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"
2 packages added:
tailor_gui tuxedo-rs (init at 0.2.1)

1 package updated:
nixos-install-tools

```

Artturin commented on pkgs/os-specific/linux/tailor-gui/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+  sourceRoot = "source/tailor_gui";
+
+  name = "tailor_gui";
```
should be pname

Artturin commented on pkgs/os-specific/linux/tailor-gui/default.nix:
```diff
@@ -0,0 +1,69 @@
+}:
+let
+
+  src = fetchFromGitHub {
+    owner = "AaronErhardt";
+    repo = "tuxedo-rs";
+    rev = "f356f05d8fe40de3c5121c11a5a9b37047501440";
+    hash = "sha256-RVauCQn8T6q+c7z0nYGtxdV/89IWAGSsqHL2xm0OTgI=";
+  };
```
src and version can be reused with `tuxedo-rs.src`

Artturin commented on pkgs/os-specific/linux/tailor-gui/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+  name = "tailor_gui";
+
+  # NOTE: OfBorg doesn't like importTOML from a subdirectory
+  version = "0.2.1";
```
That would be IFD

mrcjkb replied:
Thanks for the info.
Shall I remove the comment?

Artturin replied:
Yeah

Artturin commented on pkgs/os-specific/linux/tailor-gui/default.nix:
```diff
@@ -0,0 +1,69 @@
+    hash = "sha256-mJuCG6mPoMWinUgur+VFSx8D2WhR9KzEJsjxpzz94T8=";
+  };
+
+  nativeBuildInputs = with rustPlatform; [
+    cargoSetupHook
```
```suggestion
  nativeBuildInputs = [
    rustPlatform.cargoSetupHook
```

Artturin commented:
added platforms

mrcjkb commented:
Updated `src`.
Note: The author has not changed the version yet.

Commit:
tuxedo-rs: init at 0.2.2

Commit:
nixos/tuxedo-rs: init at 0.2.2

mrcjkb commented:
Updated from `0.2.1` -> `0.2.2`.
